README.md
# Washington DC Metro Train Sign
This project contains the source code to create your own Washington DC Metro sign. It was written using CircuitPython targeting the [Adafruit Matrix Portal](https://www.adafruit.com/product/4745) and is optimized for 64x32 RGB LED matrices.

![bd1](img/bd1.jpg)
![bd2](img/bd2.jpg)
![bd3](img/bd3.jpg)

# How To
## Hardware
- An [Adafruit Matrix Portal](https://www.adafruit.com/product/4745) - $24.99
- A **64x32 RGB LED matrix** compatible with the _Matrix Portal_ - $39.99 _to_ $84.99
    - [64x32 RGB LED Matrix - 3mm pitch](https://www.adafruit.com/product/2279)
    - [64x32 RGB LED Matrix - 4mm pitch](https://www.adafruit.com/product/2278)
    - [64x32 RGB LED Matrix - 5mm pitch](https://www.adafruit.com/product/2277)
    - [64x32 RGB LED Matrix - 6mm pitch](https://www.adafruit.com/product/2276)
- A **USB-C power supply** (15w phone adapters should work fine for this code, but the panels can theoretically pull 20w if every pixel is on white)
- A **USB-C cable** that can connect your computer/power supply to the board

## Tools
- A small phillips head screwdriver
- A hot glue gun _(optional)_
- Tape _(optional)_
- Zip ties _(optional)_

## Part 1: Prepare the Board
1. Use a hot glue gun to cover the sharp screws on the right-hand side of the 64x32 LED matrix, if present. This step is optional, but it will prevent wire chafing later on.

    ![64x32 Matrix with Hot Glue on Screws](img/base-board.jpg)

2. Lightly screw in the phillips head screws into the posts on the _Matrix Portal_. These only need to go down about 60% of the way.

    ![Matrix Portal with Screws](img/wiring.jpg)

3. Using the power cable provided with 64x32 matrix, slide the prong for the **red power cable** between the post and the screw on the port labeled **5v**. Tighten down this screw all the way using your screwdriver. Repeat the same for the **black power cable** and the **GND** port.

    ![Matrix Portal with Separate Cables](img/cables.jpg)
    ![Matrix Portal with Connected Cables](img/portal-setup.jpg)

4. Connect the _Matrix Portal_ to the large connector on the left-hand side of the back of the 64x32 matrix.

    ![64x32 Matrix with Connector Highlighted](img/port.jpg)

5. Plug one of the power connectors into the right-hand side of the 64x32 matrix.

    ![64x32 Matrix with Power Connected](img/connected-board.jpg)

6. You can use tape or zip ties to prevent the cables from flopping around.

    ![64x32 Matrix with Cable Management](img/cable-management.jpg)

## Part 2: Loading the Software
1. Connect the board to your computer using a USB-C cable. Double click the button on the board labeled _RESET_. The board should mount onto your computer as a storage volume, most likely named _MATRIXBOOT_.
    
    ![Matrix Connected via USB](img/usb-connected.jpg)

2. Flash your _Matrix Portal_ with the latest release of CircuitPython 8.
    - Download the [firmware from Adafruit](https://circuitpython.org/board/matrixportal_m4/).
    - Drag the downloaded _.uf2_ file into the root of the _MATRIXBOOT_ volume.
    - The board will automatically flash the version of CircuitPython and remount as _CIRCUITPY_.
    - If something goes wrong, refer to the [Adafruit Documentation](https://learn.adafruit.com/adafruit-matrixportal-m4/install-circuitpython).

3. Decompress the _lib.zip_ file from this repository into the root of the _CIRCUITPY_ volume. There should be one folder named _lib_, with a plethora of files underneath. You can delete _lib.zip_ from the _CIRCUITPY_ volume, as it's no longer needed.

    - It has been reported that this step may fail ([Issue #2](https://github.com/metro-sign/dc-metro/issues/2)), most likely due to the storage on the Matrix Portal not being able to handle the decompression. If this happens, unzip the _lib.zip_ file on your computer, and copy the _lib_ folder to the Matrix Portal. Command line tools could also be used if the above doesn't work.

    ![Lib Decompressed](img/lib.png)

4. Download this repository as a ZIP file by selecting the green 'Code' button at the top of this page, and then unzip the file.

5. Copy all of the Python files from the downloaded repository into the root of the _CIRCUITPY_ volume, and also copy _Metroesque.bdf_ into the _lib_ folder referred to earler.

    ![Source Files](img/source.png)

6. The board should now light up with a loading screen, but we've still got some work to do.

    ![Loading Sign](img/bd4.jpg)

## Part 3: Getting a WMATA API Key
1. Create a WMATA developer account on [WMATA's Developer Website](https://developer.wmata.com/signup/).

2. After your account is created, add the _Default Tier_ subscription to your account on [this page](https://developer.wmata.com/products/5475f1b0031f590f380924fe).

3. After doing this, you will be redirected to [your profile](https://developer.wmata.com/developer).

4. Under the _Subscriptions_ section on your profile, select the **show** button beside the _Primary Key_. This is the key that allows the board to communicate with WMATA.

## Part 4: Configuring the Board
1. Open the [config.py](src/config.py) file located in the root of the _CIRCUITPY_ volume.

2. Fill in your WiFi SSID and password under the **Network Configuration** section.

3. Under the **Metro Configuration** section:
    - Select your station and lines from the [Metro Station Codes table](#dc-metro-station-codes), and set the _metro_station_code_ value to the corresponding value in the table.
    
    - For _train_group_, the value needs to be either **'1'** or **'2'** or  **'3'**. This determines which platform's arrival times will be displayed. These typically fall in line with the values provided in the [Train Group table](#train-group-explanations), although single tracking and other events can cause these to change.
    
    - For _show_all_groups_if_nothing_else_, the value needs to be either ***True*** or ***False***. This determines what the board should do if there are no trains with assigned lines in your selected group. If True, the board will attempt to show results from all groups instead. This can help during single-tracking, when one group is temporarily reassigned to the other. If False, it will show the results from the selected group, which at most will consist of 'No Psngr' trains with no assigned lines.
    
    - Set the _metro_api_key_ value to the API key you got from [Part 3](#part-3-getting-a-wmata-api-key).

4. At the end, the first part of your configuration file should look similar this:

```python
#########################
# Network Configuration #
#########################

# WIFI Network SSID
'wifi_ssid': 'Pretty_fly_for_a_wifi',

# WIFI Password
'wifi_password': 'Panic!_at_the_cisco',

#########################
# Metro Configuration   #
#########################

# Metro Station Code
'metro_station_code': 'D02',

# Metro Train Group
'train_group': '2',

# Show All Groups If Nothing Else to Show
'show_all_groups_if_nothing_else': True,

# API Key for WMATA
'metro_api_key': 'd3adb33fd3adb33fd3adb33f',
```

5. After you save this file, your board should refresh and connect to WMATA.

## Troubleshooting
If something goes wrong, take a peek at the [Adafruit Documentation](https://learn.adafruit.com/adafruit-matrixportal-m4). Additionally, you can connect to the board using a serial connection to gain access to its logging.

# Appendix
## DC Metro Station Codes
| Name                                             | Lines      | Code |
|--------------------------------------------------|------------|------|
| Addison Road-Seat Pleasant                       | BL, SV     | G03  |
| Anacostia                                        | GR         | F06  |
| Archives-Navy Memorial-Penn Quarter              | GR, YL     | F02  |
| Arlington Cemetery                               | BL         | C06  |
| Ashburn                                          | SV         | N12  |
| Ballston-MU                                      | OR, SV     | K04  |
| Benning Road                                     | BL, SV     | G01  |
| Bethesda                                         | RD         | A09  |
| Braddock Road                                    | BL, YL     | C12  |
| Branch Ave                                       | GR         | F11  |
| Brookland-CUA                                    | RD         | B05  |
| Capitol Heights                                  | BL, SV     | G02  |
| Capitol South                                    | BL, OR, SV | D05  |
| Cheverly                                         | OR         | D11  |
| Clarendon                                        | OR, SV     | K02  |
| Cleveland Park                                   | RD         | A05  |
| College Park-U of Md                             | GR         | E09  |
| Columbia Heights                                 | GR, YL     | E04  |
| Congress Heights                                 | GR         | F07  |
| Court House                                      | OR, SV     | K01  |
| Crystal City                                     | BL, YL     | C09  |
| Deanwood                                         | OR         | D10  |
| Downtown Largo                                   | BL, SV     | G05  |
| Dunn Loring-Merrifield                           | OR         | K07  |
| Dupont Circle                                    | RD         | A03  |
| East Falls Church                                | OR, SV     | K05  |
| Eastern Market                                   | BL, OR, SV | D06  |
| Eisenhower Avenue                                | YL         | C14  |
| Farragut North                                   | RD         | A02  |
| Farragut West                                    | BL, OR, SV | C03  |
| Federal Center SW                                | BL, OR, SV | D04  |
| Federal Triangle                                 | BL, OR, SV | D01  |
| Foggy Bottom-GWU                                 | BL, OR, SV | C04  |
| Forest Glen                                      | RD         | B09  |
| Fort Totten                                      | RD         | B06  |
| Fort Totten                                      | GR, YL     | E06  |
| Franconia-Springfield                            | BL         | J03  |
| Friendship Heights                               | RD         | A08  |
| Gallery Pl-Chinatown                             | RD         | B01  |
| Gallery Pl-Chinatown                             | GR, YL     | F01  |
| Georgia Ave-Petworth                             | GR, YL     | E05  |
| Glenmont                                         | RD         | B11  |
| Greenbelt                                        | GR         | E10  |
| Greensboro                                       | SV         | N03  |
| Grosvenor-Strathmore                             | RD         | A11  |
| Herndon                                          | SV         | N08  |
| Huntington                                       | YL         | C15  |
| Hyattsville Crossing                             | GR         | E08  |
| Innovation Center                                | SV         | N09  |
| Judiciary Square                                 | RD         | B02  |
| King St-Old Town                                 | BL, YL     | C13  |
| Landover                                         | OR         | D12  |
| L'Enfant Plaza                                   | BL, OR, SV | D03  |
| L'Enfant Plaza                                   | GR, YL     | F03  |
| Loudon Gateway                                   | SV         | N11  |
| McLean                                           | SV         | N01  |
| McPherson Square                                 | BL, OR, SV | C02  |
| Medical Center                                   | RD         | A10  |
| Metro Center                                     | RD         | A01  |
| Metro Center                                     | BL, OR, SV | C01  |
| Minnesota Ave                                    | OR         | D09  |
| Morgan Boulevard                                 | BL, SV     | G04  |
| Mt Vernon Sq 7th St-Convention Center            | GR, YL     | E01  |
| Navy Yard-Ballpark                               | GR         | F05  |
| Naylor Road                                      | GR         | F09  |
| New Carrollton                                   | OR         | D13  |
| NoMa-Gallaudet U                                 | RD         | B35  |
| North Bethesda                                   | RD         | A12  |
| Pentagon                                         | BL, YL     | C07  |
| Pentagon City                                    | BL, YL     | C08  |
| Potomac Ave                                      | BL, OR, SV | D07  |
| Reston Town Center                               | SV         | N07  |
| Rhode Island Ave-Brentwood                       | RD         | B04  |
| Rockville                                        | RD         | A14  |
| Ronald Reagan Washington National Airport        | BL, YL     | C10  |
| Rosslyn                                          | BL, OR, SV | C05  |
| Shady Grove                                      | RD         | A15  |
| Shaw-Howard U                                    | GR, YL     | E02  |
| Silver Spring                                    | RD         | B08  |
| Smithsonian                                      | BL, OR, SV | D02  |
| Southern Avenue                                  | GR         | F08  |
| Spring Hill                                      | SV         | N04  |
| Stadium-Armory                                   | BL, OR, SV | D08  |
| Suitland                                         | GR         | F10  |
| Takoma                                           | RD         | B07  |
| Tenleytown-AU                                    | RD         | A07  |
| Twinbrook                                        | RD         | A13  |
| Tysons                                           | SV         | N02  |
| U Street/African-Amer Civil War Memorial/Cardozo | GR, YL     | E03  |
| Union Station                                    | RD         | B03  |
| Van Dorn Street                                  | BL         | J02  |
| Van Ness-UDC                                     | RD         | A06  |
| Vienna/Fairfax-GMU                               | OR         | K08  |
| Virginia Square-GMU                              | OR, SV     | K03  |
| Washington Dulles International Airport          | SV         | N10  |
| Waterfront                                       | GR         | F04  |
| West Falls Church                                | OR         | K06  |
| West Hyattsville                                 | GR         | E07  |
| Wheaton                                          | RD         | B10  |
| Wiehle-Reston East                               | SV         | N06  |
| Woodley Park-Zoo/Adams Morgan                    | RD         | A04  |

## Train Group Explanations
A special thanks to [u/SandBoxJohn](https://www.reddit.com/user/SandBoxJohn) for these.
| Line       | Train Group | Destination                                            |
|------------|-------------|--------------------------------------------------------|
| RD         | "1"         | Glenmont                                               |
| RD         | "2"         | Shady Grove                                            |
| BL, OR, SV | "1"         | New Carrollton, Largo Town Center                      |
| BL, OR, SV | "2"         | Vienna, Franconia-Springfield, Ashburn      |
| GR, YL     | "1"         | Greenbelt                                              |
| GR, YL     | "2"         | Huntington, Branch Avenue                              |
| N/A        | "3"         | Center Platform at National Airport, West Falls Church |


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## [1.0.0] - 2023-03-12

- Added a new font ('Metroesque.bdf') that is designed to be similar to the Metro PIDS font
- Added the number of cars in each train to the display
- Added the ability to catch more networking errors (including the dreaded "Timed out waiting for SPI char" error) and trigger an automatic reloading of _code.py_ if repeated attempts to restart the data flow are unsuccessful
- Added functionality to address single-tracking
- Added functionality to correct for inconsistencies and excessively long names in the 'Destination' reported by WMATA API (e.g., 'NewCrltn' can also be reported as 'N Carrollton' or 'NEW CARROLLTON')
- Note: The additions do not include changes from the [Scottie Garcia fork](https://github.com/scottiegarcia/dc-metro) (i.e., multiple station support, walk time filtering, optional off hours configuration, and MetroHero API Support)
- Updated CircuitPython references to CircuitPython 8 in the README
- Updated list of Metro stations in the README to reflect changed names and additional Silver Line stations

## [1.0.1] - 2023-03-18

- Updated metro_api.py with additional corrections to standardize the 'Destination' reported by the WMATA API


src/code.py
# DC Metro Board
from time import sleep
from config import config
from train_board import TrainBoard
from metro_api import MetroApi, MetroApiOnFireException
import busio
import board
from digitalio import DigitalInOut
import neopixel
from adafruit_esp32spi import adafruit_esp32spi
from adafruit_esp32spi import adafruit_esp32spi_wifimanager
from supervisor import reload

# Get wifi details and more from a secrets.py file
try:
    from secrets import secrets
except ImportError:
    print("WiFi secrets missing!")
    raise

# Set up the network connection
esp32_cs = DigitalInOut(board.ESP_CS)
esp32_ready = DigitalInOut(board.ESP_BUSY)
esp32_reset = DigitalInOut(board.ESP_RESET)
spi = busio.SPI(board.SCK, board.MOSI, board.MISO)
esp = adafruit_esp32spi.ESP_SPIcontrol(spi, esp32_cs, esp32_ready, esp32_reset)
status_light = neopixel.NeoPixel(board.NEOPIXEL, 1, brightness=0.2)
wifi = adafruit_esp32spi_wifimanager.ESPSPI_WiFiManager(esp, secrets, status_light)

if esp.status == adafruit_esp32spi.WL_IDLE_STATUS:
    print("ESP32 found and in idle mode")
print("Firmware vers.", esp.firmware_version)

print("Connecting to AP...")
while not esp.is_connected:
    try:
        esp.connect_AP(secrets['ssid'], secrets['password'])
    except OSError as e:
        print("Could not connect to AP, retrying: ", e)
        continue
print("Connected to", str(esp.ssid, 'utf-8'), "\tRSSI: ", esp.rssi)
print("Pinging google.com to establish proof of connectivity: %d ms" % esp.ping("google.com"))

station_code = config['metro_station_code']
refresh_interval = config['refresh_interval']
api = MetroApi()

# Refresh train information and deal with errors
def refresh_trains() -> [dict]:
    try:
        trains = api.fetch_train_predictions(wifi, station_code)
    except MetroApiOnFireException:
        print('WMATA API is currently on fire. Perhaps reloading will help...')
        reload()  # equivalent to using CTRL-D in a serial console
        return None
    return trains

train_board = TrainBoard(refresh_trains)

# MAIN LOOP: Call the the API every [refresh_interval] seconds
while True:
    print("Starting main loop...")
    train_board.refresh()
    sleep(refresh_interval)


src/config.py
from adafruit_bitmap_font import bitmap_font

config = {

    #########################
    # Network Configuration #
    #########################

    # WIFI Network SSID
    'wifi_ssid': '<Your 2.4ghz WiFi SSID>',
    
    # WIFI Password
    'wifi_password': '<Your wifi password>',

    #########################
    # Metro Configuration   #
    #########################

    # Metro Station Code
    # See https://github.com/GJT-34/dc-metro/blob/main/README.md for a list of stations. 
    # Arrival information for up to two stations can be displayed.
    'metro_station_code': 'A01', 

    # Metro Train Group
    # 'Group' is loosely a synonym for track.
    # See https://github.com/GJT-34/dc-metro/blob/main/README.md for a list of which groups 
    # correspond to which destinations on each line.
    'train_group': '1',

    # Show All Groups If Nothing Else to Show
    # If there are no trains with assigned lines in the selected group, we can show results for
    # all groups instead. This can help when there is single-tracking, during which time trains
    # in one group are switched to another group. 
    'show_all_groups_if_nothing_else': True,

    # API Key for WMATA
    'metro_api_key': 'd3adb33fd3adb33fd3adb33f',

    #########################
    # Other Values You      #
    # Probably Shouldn't    #
    # Touch                 #
    #########################

    'metro_api_url': 'https://api.wmata.com/StationPrediction.svc/json/GetPrediction/',
    'metro_api_retries': 2,
    'refresh_interval': 5,

    # Display Settings
    'matrix_width': 64,
    'num_trains': 3,
    'font': bitmap_font.load_font('lib/Metroesque.bdf'),
    # Metroesque font characters are GENERALLY 5px wide (including 1px for left padding) but are 
    # always 7px tall.
    #   M, m, T, V, v, W, w, & Y are 6px wide
    #   I, i, & l are 4px wide
    #   Space is 2px wide
    'character_width': 5,
    'character_height': 7,
    'bottom_text_padding': 1,

    'text_color': 0xFF5500,  # ORANGE 
    'text_color_8-car': 0x00FF00,  # GREEN
    'heading_color': 0xFF0000,  # RED
    'header': (("LN", 0, 0), ("CAR", 0, 12), ("DST", 0, 29), ("MIN", 1, 65),),

    # Size of color-coded line indicator rectangle
    'train_line_width': 3,
    'train_line_height': 7,

    # Default text loaded for each of the up-to-three "train" info lines
    'line_pos': 0,
    'loading_car_text': '-',
    'car_pos': 6,
    'loading_dest_text': 'Loading',
    'dest_pos': 13,
    'loading_min_text': '--',
    'min_pos': 65,
    'loading_line_color': 0xFF00FF,  # MAGENTA
    'dest_max_characters': 8,
}


src/train_board.py
import displayio
from adafruit_display_shapes.rect import Rect
from adafruit_display_text.label import Label
from adafruit_matrixportal.matrix import Matrix
from config import config

# Main display board that holds and updates all the pixels and data based on api data
class TrainBoard:
    """
    get_new_data is a function designed to return an array of dictionaries:
    [
        {
            'line': 0xFFFFFF,
            'car': '6',
            'destination': 'Dest Str',
            'min': '5'
        }
    ]
    """
    def __init__(self, get_new_data):
        # Function callback for API
        self.get_new_data = get_new_data

        # Object that will display our groups
        self.display = Matrix().display

        # Group that holds our objects (e.g., Rect()s and Label()s)
        self.parent_group = displayio.Group()

        # Build the heading labels
        for head in config['header']:
            self.heading = Label(
                config['font'],
                color=config['heading_color'],
                text=head[0],
                anchor_point=(head[1], 0),  # Top left (0, 0) or top right (1, 0) of label box
                anchored_position=(head[2], 0),  # x & y board location of the anchor point
                )
            self.parent_group.append(self.heading)

        # Create the main list that holds the references to our three sub-groups of train data
        # (this is not an actual displayio.Group(), just a class with data)
        self.trains = []
        for i in range(config['num_trains']):
            self.trains.append(Train(self.parent_group, i))

        # Show the board (will default to loading screen on startup)
        self.display.show(self.parent_group)

    def refresh(self) -> bool:
        train_data = self.get_new_data()

        # If we have successfully received our data from API, then update train info based on 
        # what we received
        if train_data:
            print("Reply received.")
            for i in range(config['num_trains']):
                if i < len(train_data):
                    train = train_data[i]
                    self._update_train(
                        i,
                        train['line'],
                        train['car'],
                        train['destination'],
                        train['min'],
                    )
                else:
                    self._hide_train(i)

            print("Successfully updated.")
        else:
            print("No data received. Clearing display.")

            for i in range(config['num_trains']):
                self._hide_train(i)

    def _hide_train(self, index: int):
        self.trains[index].hide()

    def _update_train(self, index: int, line:
                      int, car: str, destination: str, minutes: str):
        self.trains[index].update(line, car, destination, minutes)

# Class for holding data that defines a row of train info
class Train:
    def __init__(self, parent_group, index):
        # Get the y value of the current train section
        y = (int)(config['character_height'] + config['bottom_text_padding']) * (index + 1)

        # Build the rect that shows the line (color) of the incoming train
        self.line_rect = Rect(
            config['line_pos'], 
            y,
            config['train_line_width'],
            config['train_line_height'],
            fill=config['loading_line_color'],
        )

        self.car_label = Label(
            config['font'],
            color=config['text_color'],
            text=config['loading_car_text'],
            anchor_point=(0, 0),
            anchored_position=(config['car_pos'], y),
        )

        self.destination_label = Label(
            config['font'],
            color=config['text_color'],
            text=config['loading_dest_text'],
            anchor_point=(0, 0),
            anchored_position=(config['dest_pos'], y),
        )

        self.min_label = Label(
            config['font'],
            color=config['text_color'],
            text=config['loading_min_text'],
            anchor_point=(1, 0),
            anchored_position=(config['min_pos'], y),
        )

        self.group = displayio.Group()
        self.group.append(self.line_rect)
        self.group.append(self.car_label)
        self.group.append(self.destination_label)
        self.group.append(self.min_label)

        parent_group.append(self.group)

    def show(self):
        self.group.hidden = False

    def hide(self):
        self.group.hidden = True

    def set_line(self, line: int):
        self.line_rect.fill = line

    def set_car(self, car: str):
        car = str(car)
        self.car_label.text = car
        if car == "8":
            self.car_label.color = config['text_color_8-car']
        else:
            self.car_label.color = config['text_color']

    def set_destination(self, destination: str):
        self.destination_label.text = destination

    def set_min(self, min: str):
        minutes = str(min)
        self.min_label.text = minutes

    def update(self, line: int, car: str, destination: str, min: str):
        self.show()
        self.set_line(line)
        self.set_car(car)
        self.set_destination(destination)
        self.set_min(min)


src/metro_api.py
import time
from config import config

class MetroApiOnFireException(Exception):
    pass

class MetroApi:
    def __init__(self):
        pass

    def fetch_train_predictions(self, wifi, station_code: str) -> [dict]:
        return self._fetch_train_predictions(wifi, station_code, retry_attempt=0)

    def _fetch_train_predictions(self, wifi, station_code: str, retry_attempt: 
                                 int) -> [dict]:
        try:
            print("Fetching...")

            # Counter potential null values in API response, as python doesn't recognize null
            null = None

            # Construct a URL that will fetch data for our desired station
            api_url = config['metro_api_url'] + station_code

            # Fetch data for our desired station from the URL
            response = wifi.get(api_url, headers={'api_key': config['metro_api_key']},
                                timeout=1).json()

            # Filter train data so only objects for our selected group are in the list.
            trains = list(filter(lambda t: t['Group'] == config['train_group'], 
                          response['Trains']))
            print("Received response from API. The filtered result:")
            print(trains)
            
            # Switch to showing result for both groups if there are no trains with assigned lines 
            # in the selected group AND the user wants result from both lines in this instance. 
            # This can be useful when single-tracking occurs and one group switches to the other.
            # 'No psngr' trains with no assigned lines are ignored in figuring the group to show.            
            trains_test = list(filter(lambda t: t['Line'] in ['BL', 'GR', 'OR', 'RD', 'SV', 'YL'], 
                                      trains))
            if not trains_test and config['show_all_groups_if_nothing_else']:
                print("No trains with an assigned line in this group.")
                trains_alt = list(response['Trains'])
                trains_alt = trains_alt[:config['num_trains']]
                print("The re-filtered result, expanded to include the other one or two groups:")
                print(trains_alt)
                trains_test = list(filter(lambda t: t['Line'] in ['BL', 'GR', 'OR', 'RD', 'SV',
                                   'YL'], trains_alt))
                if not trains_test:
                    print("No trains with an assigned line in any group, so using the initial \
                          result as the basis for display.")
                else:
                    trains = list(response['Trains'])
                    print("Using the re-filtered result as the basis for display.")
                            
            # Convert train objects list to a custom list with data needed for display
            trains = [self._normalize_train_response(t) for t in trains][:config['num_trains']]
            print("Normalized train response:")
            for train in trains:
                print(train)

            return trains

        except Exception as e:
            print(e)
            if retry_attempt < config['metro_api_retries']:
                print("Failed to connect to API. Retrying...")
                # Recursion for retry logic because I don't care about your stack
                time.sleep(config['refresh_interval'])
                return self._fetch_train_predictions(wifi, station_code, 
                                                     retry_attempt + 1)
            else:
                raise MetroApiOnFireException()

    # Take a JSON object for single train and return only the data needed for display
    def _normalize_train_response(self, train: dict) -> dict:
        line = train['Line']
        line = self._get_line_color(line)

        car = train['Car']
        car = self._get_corrected_car_value(car)

        destination = train['Destination']
        destination = self._get_corrected_dest_case(destination)
        destination = self._get_corrected_dest_value(destination)

        min = train['Min']
        min = self._get_corrected_min_value(min)

        return {
            'line': line,
            'car': car,
            'destination': destination,
            'min': min
        }

    # Convert two-letter 'line" value into equivalent integer
    def _get_line_color(self, line: str) -> int:
        if line == 'BL':
            return 0x0000FF
        elif line == 'GR':
            return 0x00FF00
        elif line == 'OR':
            return 0xFF5500
        elif line == 'RD':
            return 0xFF0000
        elif line == 'SV':
            return 0xAAAAAA
        elif line == 'YL':
            return 0xFFFF00
        else:
            return 0x000000

    # Respond to empty 'car' value
    def _get_corrected_car_value(self, car: str) -> str:
        if car is None:
            car = '-'
        return car
    
    # Convert 'Destination' WMATA API result from all caps to title case
    def _get_corrected_dest_case(self, destination: str) -> str:
        # The WMATA API sometimes reports 'Destination' in all caps. Changing from all caps to 
        # titlecase takes extra work because .title() does not work in CircuitPython. 
        all_caps = 0
        dest_replace = ""
        if len(destination) > 0 and destination.isupper():
            destination = destination.lower()
            for word in destination.split():
                dest_replace += word[0].upper() + word[1:] + " "    
            dest_replace = dest_replace.rstrip()
            destination = dest_replace
        return destination
    
    # Correct for problems in 'Destination' WMATA API result
    def _get_corrected_dest_value(self, destination: str) -> str:
        # The 'Destination' reported by the WMATA API sometimes needs correcting. For instance, 
        # instead of 'NewCrltn', the API may use 'N Carrollton', which is too long to display 
        # properly. To address this, a dictionary is used to correct for known and potential 
        # problems in API-reported data. Also included in the dictionary any extra-long 
        # destinations that we don't want cropped, such as the 9-char 'Gallry Pl' (which displays
        # OK despite its length when using a variable-width font).
        # If the 1st string appears in any part of Destination, replace Destination with 2nd string
        dest_corrections = {
            'No Passenger': 'No Psngr',
            'NoPssenger': 'No Psngr',
            'ssenger': 'No Psngr',
            'Branch': 'Brnch Av',
            'Brnch Ave': 'Brnch Av',
            'Carroll': 'NewCrltn',
            'Chinatown': 'Gallry Pl',
            'Downtown': 'Largo',
            'DT Largo': 'Largo',
            'Dulles': 'Dulles',
            'Fairfax': 'Fairfax',
            'Farragut N': 'Frgut N.',
            'Fran/': 'Frnconia',
            'Franconia': 'Frnconia',
            'Ft. Tottn': 'Ft.Tottn',
            'Gallery': 'Gallry Pl',
            'Gallry Pl': 'Gallry Pl',
            'GallryPl': 'Gallry Pl',
            'Greenbelt': 'Grnbelt',
            'Grosvnor': 'Grsvnor',
            'Huntington': 'Hntingtn',
            'NoMa': 'NoMa',
            'Noma': 'NoMa',
            'R.i.': 'R.I.',
            'Rhode': 'R.I.',
            'Shady': 'Shady Gr',
            'Silver': 'SilvrSpr',
            'Silvrspr': 'SilvrSpr',
            'Totten': 'Ft.Tottn',
            'U St': 'U St',
            'Vernon': 'Mt. Vern',
            'W Falls': 'W Fls Ch',
            'West Falls': 'W Fls Ch'
        }
        dest_replace = ""
        for key in dest_corrections.keys():
            if key in destination:
                dest_replace = dest_corrections[key]
        if dest_replace:
            destination = dest_replace
        else:
            destination = destination[:config['dest_max_characters']]
        return destination
    
    # Reduce three-character 'min' values to two characters, to save space on the train board
    def _get_corrected_min_value(self, min: str) -> str:
        if min == 'ARR':
            min = 'AR'
        if min == 'BRD':
            min = 'BD'
        if min == '---' or min == '':
            min = 'DL'
        return min
